database/engine/rrdenginerepl.c

```diff
@@ -0,0 +1,246 @@
+        }
+    }
+    // page flags check. Need to enable the pg_cache_descr_state flags
+    pg_cache_insert(ctx, page_index, descr);
+    // Try to update the time start and end of the metric.
+    pg_cache_add_new_metric_time(page_index, descr);
+
+    debug(D_REPLICATION, "Commiting page in dbengine....");
+    if (likely(descr->page_length)) {
+        int page_is_empty;
+
+        rrd_stat_atomic_add(&ctx->stats.metric_API_producers, -1);
+
+        page_is_empty = rrdeng_page_has_only_empty_metrics(descr);
+        if (page_is_empty) {
+            error("Past metrics page has empty metrics only, deleting:");
+            pg_cache_put(ctx, descr);
+            pg_cache_punch_hole(ctx, descr, 1, 0, NULL);
+        } else {
+            rrdeng_commit_page(ctx, descr, dim_past_data->page_correlation_id);
+        }
+    } else {
+        freez(descr->pg_cache_descr->page);
+        rrdeng_destroy_pg_cache_descr(ctx, descr->pg_cache_descr);
+        freez(descr);
+    }
+    debug(D_REPLICATION, "Page Commited -  Dimension \"%s\".\"%s\" metrics page commited in memory.", rd->rrdset->id, rd->id);
+    // TBR: Only for debug
+    debug(D_REPLICATION, "REP: OBSERVE dimension (%s.%s) in time_interval[%ld, %ld] #samples(%lu)....END", rd->rrdset->id, rd->id, (time_t)(descr->start_time/USEC_PER_SEC), (time_t)(descr->end_time/USEC_PER_SEC), (descr->page_length/sizeof(storage_number)));
+}
+
+void rrdeng_store_past_metrics_page_finalize(RRDDIM_PAST_DATA *dim_past_data){
+    struct pg_cache_page_index* page_index;
+    RRDDIM *rd = dim_past_data->rd;
+    page_index = dim_past_data->rd->state->page_index;
+#ifndef NETDATA_INTERNAL_CHECKS
+    UNUSED(rd);
+#endif
+
+    uv_rwlock_wrlock(&page_index->lock);
+    --page_index->writers;
+    uv_rwlock_wrunlock(&page_index->lock);
+    debug(D_REPLICATION, "Finalize operation -  Dimension \"%s\".\"%s\" metrics page completed.", rd->rrdset->id, rd->id);
+}
+
+// It saves the GAP past metrics in the active page in real-time
+int rrdeng_store_past_metrics_realtime(RRDDIM *rd, RRDDIM_PAST_DATA *dim_past_data)
+{
+    struct rrdeng_collect_handle *handle = (struct rrdeng_collect_handle *)rd->state->handle;
+    struct rrdeng_page_descr *descr;
+    storage_number *page, *page_gap;
+    int return_value = 0;
+
+    descr = handle->descr;
+    if(!descr || !descr->pg_cache_descr) {
+        infoerr("No active descr or page for dimension %s.%s", rd->rrdset->id, rd->id);
+        return 1;
+    }
+
+    page = (storage_number *)descr->pg_cache_descr->page;
+    page_gap = (storage_number *)dim_past_data->page;
+
+    uint64_t start, end, page_start, page_end;
+    start = dim_past_data->start_time / USEC_PER_SEC; //gap time start
+    end = dim_past_data->end_time / USEC_PER_SEC;     //gap time end
+    page_start = descr->start_time / USEC_PER_SEC;    //active page time start
+    page_end = descr->end_time / USEC_PER_SEC;        //active page time start
+
+    if (!page || !page_gap || start > end) {
```

Comment:
You can do these tests before all the divisions that happens between lines `180-184`.

Reply:
Not really because the error message shows the computed information.